Remove Order Fulfillment Screen

shiki opened the pull request:
Part of #3034. Closes #1619. Closes #1906. 

This removes the order fulfillment screen and moves the Mark Order Complete button on the Order Details page. This was done on Android a while ago. 

🙇 Sorry for the long PR, I just passed 500 lines and I don't know how I could make this even shorter. 

## Design

### Before

https://user-images.githubusercontent.com/198826/104258305-7e612a00-543c-11eb-9871-1489b7c265bf.mp4       

### After

https://user-images.githubusercontent.com/198826/104258087-19a5cf80-543c-11eb-9b39-6d17cd8d5835.mp4 

### Error Messages

Fulfillment Failed | Undo Fulfillment Failed
--------|-------
![Simulator Screen Shot - iPhone 8 - 2021-01-11 at 19 05 43](https://user-images.githubusercontent.com/198826/104260745-51fbdc80-5441-11eb-963a-deeff308251e.png)|  ![Simulator Screen Shot - iPhone 8 - 2021-01-11 at 19 13 44](https://user-images.githubusercontent.com/198826/104260751-54f6cd00-5441-11eb-894f-0cac161b57c2.png)     
  






## Changes

Most of the fulfillment logic in here was copied from [`FulfillViewController.fulfillWasPressed`](https://github.com/woocommerce/woocommerce-ios/blob/e18681e70bb2a77cefee885a66bdc5a19371ca82/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/Product%20List%20Section/Fulfillment/FulfillViewController.swift#L179-L207). To make things a bit more testable and encapsulated, I split them out into two parts:

- [`OrderFulfillmentUseCase`](https://github.com/woocommerce/woocommerce-ios/blob/issue/3034-remove-order-fulfillment/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/OrderFulfillmentUseCase.swift). This class has the business logic of fulfilling, retrying, or undoing the status changes. 
- [`OrderFulfillmentNoticePresenter`](https://github.com/woocommerce/woocommerce-ios/blob/issue/3034-remove-order-fulfillment/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/OrderFulfillmentNoticePresenter.swift). This class is the UI handler for `OrderFulfillmentUseCase`. It presents the notices for informing the user that the fulfillment was completed or failed. 

Please see the documentation in the classes for more information. 

## Next Steps

While this fully replaces the Fulfillment screen, I still have more to do: 

- Remove the old `FulfillViewController` code and xib. 
- Add unit tests for `OrderFulfillmentNoticePresenter`. 

I'll do these in follow-up PRs. 

## Testing

### Normal Operation

1. Navigate to a Processing order. 
2. Tap on Mark Order Complete. 
3. Confirm in WC Admin that the order was changed to Completed.
4. Navigate to another Processing order. 
5. Tap on Mark Order Complete and then quickly tap the Undo button. 
6. Confirm in WC Admin that the order was **not** changed to Completed.

### Offline

1. Go offline. 
2. Navigate to a Processing order. 
3. Tap on Mark Order Complete. 
4. Confirm that a notice “Unable to fulfill order...” was eventually shown. 
5. Tap on the Retry button on that notice. 
6. Confirm that a notice “Unable to fulfill order...” was shown. 

### Offline - Retry

1. Go offline. 
2. Navigate to a Processing order. 
3. Tap on Mark Order Complete. 
4. Go online while waiting for the “Unable to fulfill order...” notice gets displayed.
5. Tap on Retry.
6. Confirm in WC Admin that the order was changed to Completed.

## Submitter Checklist

- [x] If it's feasible, I have added unit tests. 
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

shiki committed:
Add "Mark Order Complete" button in Order Details

shiki committed:
Fix typo

shiki committed:
OrderDetailsVM: Inject StoresManager

shiki committed:
Remove duplicate updateOrderStatus(order:)

This just does the same as update(order:) so we might as well just use one of them.

shiki committed:
Add UseCase for marking an order as complete

shiki committed:
Add accompanying Presenter for OrderFulfillmentUseCase

shiki committed:
Handle Mark Order as Completed button on Order Details

shiki committed:
Hide Mark Order as Complete button behind feature flag

shiki committed:
Update naming of test method to match current standard

shiki committed:
OrderDetailsVMTests: Simplify import

shiki committed:
OrderDetailsVMTests: Use mock StoresManager

shiki committed:
OrderDetailsVMTests: Add Given/When/Then

shiki committed:
Move OrderFulfillmentUseCase to inside ViewRelated

shiki committed:
Add test for OrderDetailsViewModel.markComplete

shiki committed:
Rename FulfillmentProcess.future to result

shiki committed:
Document OrderFulfillmentUseCase

shiki committed:
Turn on removeOrderFulfillmentScreen flag for developer

shiki committed:
Document OrderFulfillmentNoticePresenter

shiki committed:
Add initial OrderFulfillmentUUseCaseTests

shiki committed:
OrderFulfillmentUseCase: Add unit test for undo

shiki committed:
OrderFulfillmentUseCase: Add test for retry

shiki committed:
OrderFulfillmentUseCase: Add tests for result Future completion

shiki committed:
Add missing word in doc

shiki committed:
Remove fulfillment screen feature flag

I think I don't need it after all.

shiki committed:
Simplify doc of OrderFulfillmentNoticePresenter

shiki committed:
Remove unused featureFlagService

shiki committed:
Remove unused import

shiki committed:
Update RELEASE NOTES

shiki committed:
Merge branch 'develop' into issue/3034-remove-order-fulfillment

shiki commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift:
```diff
@@ -128,10 +130,6 @@ extension OrderDetailsViewModel {
     func configureResultsControllers(onReload: @escaping () -> Void) {
         dataSource.configureResultsControllers(onReload: onReload)
     }
-
-    func updateOrderStatus(order: Order) {
-        update(order: order)
-    }
```
I removed this because it was just the same as `update(order:)`. 🤷

jaclync commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift:
```diff
@@ -1174,6 +1184,7 @@ extension OrderDetailsDataSource {
         case summary
         case aggregateOrderItem
         case fulfillButton
```
nit: we can remove this case now that it's replaced by case `markCompleteButton`?

shiki replied:
I will remove all of these in an upcoming PR. 🙂

jaclync commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift:
```diff
@@ -1256,6 +1269,7 @@ extension OrderDetailsDataSource {
 
     enum CellActionType {
         case fulfill
```
nit: similar to above, this case could be removed?

shiki replied:
I made a note to remove this in an upcoming PR which will remove the `FullfillViewController` and everything else related to it.

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderFulfillmentNoticePresenter.swift:
```diff
@@ -0,0 +1,70 @@
+///
+final class OrderFulfillmentNoticePresenter {
+
+    private let noticePresenter: NoticePresenter = ServiceLocator.noticePresenter
+    private let analytics: Analytics = ServiceLocator.analytics
```
nit: maybe these two dependencies can be DI'ed in the initializer for testing later?

shiki replied:
That's the plan. I'll add the DI when I add the unit tests. I still have to figure out how I'm gonna unit test Notices in the first place. 😬

jaclync commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderFulfillmentNoticePresenter.swift:
```diff
@@ -0,0 +1,70 @@
+                self.displayFulfillmentErrorNotice(error: fulfillmentError)
+            }
+
+            cancellable.cancel()
```
just curious, does a subscription automatically end when an observable is completed?

if so, I wonder if this line isn't needed and we can just call `.store(in: &cancellables)` in L39?

shiki replied:
Oh, good point. The `cancel()` is not necessary I think because `Future` is like `Single`. When it completes, it doesn't send any values anymore. So the `Subscription` effectively ends. 

The important thing is to remove the `AnyCancellable` from memory. Based on my test, the `Subscriber.Sink` and `AnyCancellable` were still kept in memory because they're stored in `self.cancellables` (`Set<>`). I added a removal in f0bd0e50d6070cdea2d2b400021214e1df7ba4d1. 

Thanks for pointing that out!

I also noticed the `OrderFulfillmentNoticePresenter` still kept in memory but I can't figure out why. I'll fix it in another PR.

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Details/OrderDetailsDataSourceTests.swift:
```diff
@@ -83,6 +83,34 @@ final class OrderDetailsDataSourceTests: XCTestCase {
+        XCTAssertNotNil(buttonRow)
+    }
+
+    func test_markOrderComplete_button_is_hidden_if_order_is_not_processing() throws {
```
nice tests!

jaclync commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Details/OrderFulfillmentUseCaseTests.swift:
```diff
@@ -0,0 +1,185 @@
+@testable import WooCommerce
+
+/// Unit tests for `OrderFulfillmentUseCase`.
+final class OrderFulfillmentUseCaseTests: XCTestCase {
```
wow very detailed tests 💯

shiki replied:
Thanks. 🙂 I hoped that these can be shorter. I wonder if Combine has some unit-testing utilities. 🤔 Kotlin has those for Coroutines. 

https://github.com/woocommerce/woocommerce-ios/blob/f86b69718679f3e95a646fdc7a1e0f2945a80665/WooCommerce/WooCommerceTests/ViewRelated/Orders/Order%20Details/OrderFulfillmentUseCaseTests.swift#L81-L90

jaclync reviewed (approved):
Nice work separating the use case and presenter with very thorough unit tests!

Code: ✅ 
Testing: ✅ went through all 3 test scenarios and worked as expected!

Very minor and non-blocking observations, no actions needed:
- In the offline scenario, the "Mark Order Complete" CTA reappeared instantly while the "Order marked as fulfilled" is shown, then followed by the error notice. The second error notice is probably delayed because "Order marked as fulfilled" notice is shown for like 1s based on the `DefaultNoticePresenter` implementation 🤷🏻‍♀️
- The order number is formatted with a thousand separator for some reason (e.g. `Failed to undo fulfillment of order #3,285`):
<img src="https://user-images.githubusercontent.com/1945542/104296085-330e5080-54fc-11eb-9f96-6766a8b77325.png" width="300" />

shiki committed:
Remove the cancellable when the Future completes

It doesn't look like we need to call cancel() because AnyCancellable should be
cancelled automatically when it is deinitialized. The important thing is to
remove it from the self.cancellables list so **it is** deinitialized.

shiki reviewed (commented):
Thanks, Adam and @jaclync! 

> In the offline scenario, the "Mark Order Complete" CTA reappeared instantly while the "Order marked as fulfilled" is shown, then followed by the error notice. The second error notice is probably delayed because "Order marked as fulfilled" notice is shown for like 1s based on the DefaultNoticePresenter implementation 🤷🏻‍♀️

I think what we can do here is cancel the Notice when the order failed. We still have that functionality as a debt here: https://github.com/woocommerce/woocommerce-ios/issues/1752. 

> The order number is formatted with a thousand separator for some reason (e.g. Failed to undo fulfillment of order #3,285):

Good catch! I only copied the original code but I'll fix it in a separate PR.
